internal/metrics/benchmark_test.go

```diff
@@ -0,0 +1,108 @@
+// Copyright (c) 2024 The Jaeger Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+package benchmark_test
+
+import (
+	"context"
+	"testing"
+
+	promsdk "github.com/prometheus/client_golang/prometheus"
+	"github.com/stretchr/testify/require"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/bridge/opencensus"
+	"go.opentelemetry.io/otel/metric"
+	sdkmetric "go.opentelemetry.io/otel/sdk/metric"
+
+	promExporter "go.opentelemetry.io/otel/exporters/prometheus"
+)
+
+func BenchmarkPrometheusCounter(b *testing.B) {
+	reg := promsdk.NewRegistry()
+	opts := promsdk.CounterOpts{
+		Name: "test_counter",
+		Help: "help",
+	}
+	cv := promsdk.NewCounterVec(opts, []string{"tag1"})
+	reg.MustRegister(cv)
+	counter := cv.WithLabelValues("value1")
+
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		counter.Add(1)
+	}
+}
+
+func BenchmarkOTELCounter(b *testing.B) {
+	counter := otelCounter(b)
+	ctx := context.Background()
+
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		counter.Add(ctx, 1)
+	}
+}
+
+func BenchmarkCencusCounter(b *testing.B) {
+	counter := censusCounter(b)
+	ctx := context.Background()
+
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		counter.Add(ctx, 1)
+	}
+}
+
+func BenchmarkOTELCounterWithLabel(b *testing.B) {
+	counter := otelCounter(b)
+	ctx := context.Background()
+	attrSet := attribute.NewSet(attribute.String("tag1", "value1"))
+	attrOpt := metric.WithAttributeSet(attrSet)
+
```

Comment:
Curious how you feel about these two lines, and why you've excluded them from the benchmark? Ergonomically speaking, would you cache the `attrOpt` value after computing it? If so, you'd be better off with bound instruments. If not, you should measure it. I ask because the introduction of functional options adds a bunch of allocations, so unless you compute and re-use the []Option slice, you've either got an ergonomics problem or a performance problem. We stopped using the OTel-Go API as a result of this and installed a more-efficient functional-option-free bypass. https://github.com/lightstep/otel-launcher-go/issues/446

Reply:
> Ergonomically speaking, would you cache the attrOpt value after computing it?

@jmacd yes, that's exactly what Jaeger is doing, we always used "bound" instruments. https://github.com/jaegertracing/jaeger/blob/8a338003d460ce4dabe425b631f96d08e6ee01b2/internal/metrics/otelmetrics/counter.go#L12-L20

But this does not help to completely avoid allocations. It's not the passing of vararg options that's causing the allocations, it's something deeper in the implementation.